plutus-core/cost-model/create-cost-model/CreateBuiltinCostModel.hs

```diff
@@ -235,241 +228,217 @@ createBuiltinCostModel bmfile rfile = do
+    "linear_in_z"                 -> ModelThreeArgumentsLinearInZ             <$> readOneVariableLinearFunction "z_mem" e
+    "quadratic_in_z"              -> ModelThreeArgumentsQuadraticInZ          <$> readOneVariableQuadraticFunction "z_mem" e
+    "literal_in_y_or_linear_in_z" -> ModelThreeArgumentsLiteralInYOrLinearInZ <$> error "literal"
+    _                             -> error $ "Unknown three-variable model type: " ++ ty
+
+loadSixVariableCostingFunction :: MonadR m => SomeSEXP (Region m) -> m ModelSixArguments
+loadSixVariableCostingFunction e = do
+  ty <- getType e
+  case ty of
+    "constant_cost" -> ModelSixArgumentsConstantCost <$> getConstant e
+    _               -> error $ "Unknown six-variable model type: " ++ ty
 
 boolMemModel :: ModelTwoArguments
 boolMemModel = ModelTwoArgumentsConstantCost 1
 
 memoryUsageAsCostingInteger :: ExMemoryUsage a => a -> CostingInteger
 memoryUsageAsCostingInteger = coerce . sumCostStream . flattenCostRose . memoryUsage
 
+-- | The types of functions which take an R SEXP and extract a CPU model from
+-- it, then pair it up with a memory costing function.
+type MakeCostingFun1 = forall m . MonadR m => SomeSEXP (Region m) -> m (CostingFun ModelOneArgument)
+type MakeCostingFun2 = forall m . MonadR m => SomeSEXP (Region m) -> m (CostingFun ModelTwoArguments)
+type MakeCostingFun3 = forall m . MonadR m => SomeSEXP (Region m) -> m (CostingFun ModelThreeArguments)
+type MakeCostingFun6 = forall m . MonadR m => SomeSEXP (Region m) -> m (CostingFun ModelSixArguments)
+
+-- | Create a
+makeOneVariableCostingFunction :: ModelOneArgument -> MakeCostingFun1
+makeOneVariableCostingFunction memModel e =
+  CostingFun <$> (loadOneVariableCostingFunction e) <*> pure memModel
+
+makeTwoVariableCostingFunction :: ModelTwoArguments -> MakeCostingFun2
+makeTwoVariableCostingFunction memModel e =
+  CostingFun <$> (loadTwoVariableCostingFunction e) <*> pure memModel
+
+makeThreeVariableCostingFunction :: ModelThreeArguments -> MakeCostingFun3
+makeThreeVariableCostingFunction memModel e =
+  CostingFun <$> (loadThreeVariableCostingFunction e) <*> pure memModel
+
+makeSixVariableCostingFunction :: ModelSixArguments -> MakeCostingFun6
+makeSixVariableCostingFunction memModel e =
+  CostingFun <$> (loadSixVariableCostingFunction e) <*> pure memModel
 
 ---------------- Integers ----------------
 
-addInteger :: MonadR m => SomeSEXP (Region m) -> m (CostingFun ModelTwoArguments)
-addInteger cpuModelR = do
-  cpuModel <- ModelTwoArgumentsMaxSize <$> readModelMaxSize cpuModelR
-  -- The worst case is adding e.g. `maxBound :: Int` + `maxBound :: Int`, which increases the memory usage by one.
-  -- (max x y) + 1
-  let memModel = ModelTwoArgumentsMaxSize $ ModelMaxSize 1 1
-  pure $ CostingFun cpuModel memModel
+addInteger :: MakeCostingFun2
+addInteger =
+  makeTwoVariableCostingFunction $ ModelTwoArgumentsMaxSize $ OneVariableLinearFunction 1 1
```

Comment:
So we unhardcode parsing, but we still have to keep those models hardcoded here, because that's how we get the data and there's no other way we could infer them, right?

Reply:
That's right. If we wanted to be totally general I suppose we could return the model formulas from R (things like [this](https://github.com/IntersectMBO/plutus/blob/a79131eecfec1292b37eb2cce1e6299b46f360cc/plutus-core/cost-model/data/models.R#L424)) and try to parse them and then we could deal with more or less arbitrary functions (and we'd also need some language for describing regions of the input space where different cost models apply).  That'd be the most generic solution to PLT-408, but it might be overkill for what we're doing.

Reply:
Yeah, that does sound like an overkill. Thanks!

It would be really great if we could somehow have these models hardcoded here and reused within the definitions of builtins. No JSON for shapes, just taking the shapes directly from this place and using them where we need them. That would be the most reliable and ergonomic solution I think, but for that we also need to untangle shapes from constants (currently we put constants right into the shapes basically).

I feel like my approach would be better long-term. Short-term, not so sure, perhaps yours is better. Both are complicated though: with yours we need to have multiple JSON files or to add support for some form of versioning into the existing JSON file (right?) and with mine we need to rewrite half the costing code. Yours sounds annoying, mine sounds insufferable.